pkg/kubelet/kubelet.go

```diff
@@ -678,20 +680,83 @@ func (kl *Kubelet) runOOMWatcher() {
 	}
 }
 
+func (kl *Kubelet) initialNodeStatus() (*api.Node, error) {
+	node := &api.Node{
+		ObjectMeta: api.ObjectMeta{
+			Name:   kl.hostname,
+			Labels: map[string]string{"kubernetes.io/hostname": kl.hostname},
+		},
+	}
+	if kl.cloud != nil {
+		instances, ok := kl.cloud.Instances()
+		if !ok {
+			return nil, fmt.Errorf("failed to get instances from cloud provider")
+		}
+		// TODO(roberthbailey): Can we do this without having credentials to talk
+		// to the cloud provider?
+		instanceID, err := instances.ExternalID(kl.hostname)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get instance ID from cloud provider: %v", err)
+		}
+		node.Spec.ExternalID = instanceID
+	} else {
+		node.Spec.ExternalID = kl.hostname
+	}
+	if err := kl.setNodeStatus(node); err != nil {
+		return nil, err
+	}
+	return node, nil
+}
+
+// registerWithApiserver registers the node with the cluster master.
+func (kl *Kubelet) registerWithApiserver() {
+	step := 100 * time.Millisecond
+	for {
+		time.Sleep(step)
+		step = step * 2
+		if step >= 7*time.Second {
+			step = 7 * time.Second
+		}
+
+		node, err := kl.initialNodeStatus()
+		if err != nil {
+			glog.Errorf("Unable to construct api.Node object for kubelet: %v", err)
+			continue
+		}
+		glog.V(2).Infof("Attempting to register node %s", node.Name)
+		if _, err := kl.kubeClient.Nodes().Create(node); err != nil {
+			if apierrors.IsAlreadyExists(err) {
+				currentNode, err := kl.kubeClient.Nodes().Get(kl.hostname)
+				if err != nil {
+					glog.Errorf("error getting node %q: %v", kl.hostname, err)
+					continue
+				}
+				if currentNode == nil {
+					glog.Errorf("no node instance returned for %q", kl.hostname)
+					continue
+				}
+				if currentNode.Spec.ExternalID == node.Spec.ExternalID {
```

Comment:
What does it mean if we get here and these are != ? Should we delete and re-register the node instead of just returning?